packages/html-mod/src/experimental/index.ts

```diff
@@ -0,0 +1,1125 @@
+import {
+  parseDocument,
+  SourceDocument,
+  SourceElement,
+  Options,
+  DomUtils,
+  nodeToString,
+  SourceText,
+} from '@ciolabs/htmlparser2-source';
+import { select } from 'cheerio-select';
+import escapeHtml from 'escape-html';
+import { decode } from 'html-entities';
+import MagicString from 'magic-string';
+
+import * as AstManipulator from './ast-manipulator.js';
+import { AstUpdater } from './ast-updater.js';
+import {
+  calculateOverwriteDelta,
+  calculateAppendRightDelta,
+  calculatePrependLeftDelta,
+  calculateRemoveDelta,
+  type PositionDelta,
+} from './position-delta.js';
+
+export type HtmlModOptions = Options & {
+  HtmlModElement?: typeof HtmlModElement; // allow for custom HtmlModElement
+};
+
+export class HtmlMod {
+  __source: string;
+  __s: MagicString;
+  __dom: SourceDocument;
+  __flushed = false;
+  __HtmlMod: typeof HtmlMod;
+  __HtmlModElement: typeof HtmlModElement;
+  __HtmlModText: typeof HtmlModText;
+  __options: HtmlModOptions;
+  __astUpdater: AstUpdater;
+  __pendingDeltas: PositionDelta[] = [];
+  __cachedInnerHTML: WeakMap<SourceElement, string> = new WeakMap();
+  __cachedOuterHTML: WeakMap<SourceElement, string> = new WeakMap();
+
+  constructor(source: string, options?: HtmlModOptions) {
+    this.__source = source;
+    this.__options = {
+      recognizeSelfClosing: true,
+      ...options,
+    };
+    this.__s = new MagicString(source);
+    this.__dom = parseDocument(source, this.__options);
+    this.__flushed = true;
+    this.__HtmlModElement = options?.HtmlModElement || HtmlModElement;
+    this.__HtmlModText = HtmlModText;
+    this.__HtmlMod = HtmlMod;
+    this.__astUpdater = new AstUpdater();
+  }
+
+  /**
+   * Apply all pending deltas at end of operation
+   * This updates AST positions, refreshes source, and creates fresh MagicString
+   */
+  __finishOperation() {
```

Comment:
What if we only finish the operation on the reads? So when we do bulk manipulation we only flush the deltas once l?